tyrian/js/src/main/scala/tyrian/websocket/WebSocket.scala

```diff
@@ -2,22 +2,23 @@ package tyrian.websocket
 
 import cats.effect.kernel.Async
 import cats.effect.kernel.Sync
-import cats.syntax.functor.*
+import cats.effect.std.Dispatcher
+import cats.syntax.all.*
 import fs2.Stream
+import fs2.concurrent.Channel
 import org.scalajs.dom
 import tyrian.Cmd
 import tyrian.Sub
 import tyrian.websocket.WebSocketEvent
 import util.Functions
 
-import scala.collection.mutable.Queue
 import scala.concurrent.duration.*
 
 /** Helper WebSocket instance to store in your model that acts as a controller */
 final class WebSocket[F[_]: Async](liveSocket: LiveSocket[F]):
   /** Disconnect from this WebSocket */
   def disconnect[Msg]: Cmd[F, Msg] =
-    Cmd.SideEffect(liveSocket.socket.close(1000, "Graceful shutdown"))
+    Cmd.SideEffect(Async[F].delay(liveSocket.socket.close(1000, "Graceful shutdown")) *> liveSocket.closeChannel)
```

Comment:
I think this could be `guarantee` instead of `*>`?

Reply:
> if the websocket didn't close, should we close the dispatcher and channel?

I was thinking more about the case where the network call fails, so we don't get the close acknowledgement from the server, but our connection gets closed regardless (think closing a tab). I think the right thing to do is to guarantee the channel is closed as well, after all we are specifically calling `disconnect`, but I see how that can get philosophical 😄

Reply:
> I was thinking more about the case where the network call fails

Right, this is an excellent point 🤔 

> I think the right thing to do is to guarantee the channel is closed as well

Can you talk me through exactly how that would work in the case of network failure? I can't see how adding a guarantee here would guard against network failure.

I have a really hard time reasoning about resource safety in Tyrian :)

Reply:
> I can't see how adding a guarantee here would guard against network failure.

We wouldn't guard against a network failure. We can let it fail while ensuring (via `guarantee`) that the resources are properly closed to avoid potential leaks (channel, and dispatcher that uses allocated).

If that's not convincing enough, feel free to ignore and leave it as is, hardly any free time right now, sorry 😄

Reply:
Thanks!

> We can let it fail while ensuring (via `guarantee`) that the resources are properly closed to avoid potential leaks

Yes, absolutely, :100: agree. We need to handle the errors where they are raised, which is actually not here. See https://github.com/PurpleKingdomGames/tyrian/pull/177#discussion_r1141033595.

Reply:
So my reading of this then is that you're going to skip `guarantee` because you've done the 'close on error' work, which seems to be the more correct place to deal with this sort of thing. Have I got that right?

Reply:
Yes, I don't think that `guarantee` makes sense here. From my reading of MDN, `websocket.close()` will only raise an exception if the socket actually didn't close (due to invalid args which should never happen anyway). In which case my opinion is we _don't_ want to close the channel or the dispatcher, if the socket itself is not closed.

https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close#exceptions

Taking a step back: this actually makes sense. `websocket.close()` is a synchronous call, so it can only raise exceptions for synchronous errors. If there is some sort of network failure, this occurs asynchronously, and would not be raised here.